onnxruntime/core/optimizer/qdq_transformer/selectors_actions/qdq_actions.h

```diff
@@ -76,6 +80,24 @@ struct MatMulReplaceWithQLinear : public Action {
   BinaryReplaceWithQLinear qlinear_matmul_replacer_;
 };
 
+// used together with DQMatMulNodeGroupSelector, which does the sanity check
+struct DQMatMulReplaceWithMatMulNBits : public Action {
+  explicit DQMatMulReplaceWithMatMulNBits(int64_t accuracy_level);
+  Status Run(Graph&, const NodesToOptimize& selected_nodes) const override;
```

Comment:
if you want to support use with runtime optimizations in an ORT format model, probably need to implement `RunForSave()` as well.

it should record the newly added op's schema. e.g.,
https://github.com/microsoft/onnxruntime/blob/8c2689877fb48bfb4a6a133b020cdb5ec7c9b066/onnxruntime/core/optimizer/selectors_actions/actions.cc#L115-L135

it may be simpler to reuse `ReplaceWithNew` if possible.

Reply:
`ReplaceWithNew` does not support adding new initializers to the graph. It supports the scenario of replacing nodes and moving input/outputs.

Reply:
We're still replacing nodes with a new one and moving input/outputs aren't we, so re-using ReplaceWithNew seems better than re-implementing that aspect of the optimization?

We could add a class derived from ReplaceWithNew that supports adding initializers, or if it doesn't break any existing usage the current ReplaceWithNew implementation couldn't be expanded to support adding initializers.

Reply:
fix in next iter